scheduler/scheduler.go

```diff
@@ -69,17 +82,113 @@ type Queue struct {
 	queuedRequestsMetricsKey string
 }
 
+// FifoQueue objects implement a FIFO queue for handling requests,
+// with a maximum allowed concurrency and queue size. Currently, they
+// can be used from the fifo filters in the filters/scheduler package
+// only.
+type FifoQueue struct {
+	queue                    *fifoQueue
+	config                   Config
+	metrics                  metrics.Metrics
+	activeRequestsMetricsKey string
+	errorFullMetricsKey      string
+	errorOtherMetricsKey     string
+	errorTimeoutMetricsKey   string
+	queuedRequestsMetricsKey string
+}
+
+type fifoQueue struct {
+	mu             sync.RWMutex
+	counter        *atomic.Uint64
+	sem            *semaphore.Weighted
+	timeout        time.Duration
+	maxQueueSize   uint64
+	maxConcurrency uint64
+	closed         bool
+}
+
+func (fq *fifoQueue) Status() QueueStatus {
+	fq.mu.RLock()
+	defer fq.mu.RUnlock()
+	maxConcurrency := fq.maxConcurrency
+	all := fq.counter.Load()
+	queued := float64(0)
+	if all > maxConcurrency {
+		queued = math.Max(float64(all-maxConcurrency), 0)
+	}
+	active := math.Min(float64(all), float64(maxConcurrency))
+	return QueueStatus{
+		ActiveRequests: int(active),
+		QueuedRequests: int(queued),
+		Closed:         fq.closed,
+	}
+}
+
+// Reconfigure does nothing, because it creates a data race that would
+// require more mutex than we want to have
+func (fq *fifoQueue) Reconfigure(c Config) {
+	fq.mu.Lock()
+	defer fq.mu.Unlock()
+	fq.maxConcurrency = uint64(c.MaxConcurrency)
+	fq.maxQueueSize = uint64(c.MaxQueueSize)
+	fq.timeout = c.Timeout
+}
+
+func (fq *fifoQueue) Wait(ctx context.Context) (func(), error) {
+	fq.mu.RLock()
+	maxConcurrency := fq.maxConcurrency
+	maxQueueSize := fq.maxQueueSize
+	timeout := fq.timeout
+	fq.mu.RUnlock()
+
+	// queue full?
+	if fq.counter.Load() > maxConcurrency+maxQueueSize {
+		return func() {}, ErrQueueFull
+	}
+
+	// handle queue
+	fq.counter.Inc()
+	defer func() {
+		fq.counter.Dec()
+	}()
```

Comment:
I think here we can do
```go
n := fq.counter.Inc()
defer fq.counter.Dec()

if n > maxConcurrency+maxQueueSize {
  return nil, ErrQueueFull
}
```
without using Load()

Reply:
thanks true, fixed